torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1167,60 +856,335 @@ def _get_ignored_params(
-        excluding any :class:`FlatParameter` s and their fully prefixed names,
+        excluding any :class:`FlatParameter` s, and their fully prefixed names,
         both as :class:`set` s.
-
-        Args:
-            root_module (torch.nn.Module): Top-level module passed into the
-                FSDP constructor from which to derive the fully prefixed names.
-            ignored_modules (Set[torch.nn.Module]): Modules to ignore.
         """
         ignored_params = set(
-            p for m in ignored_modules for p in m.parameters()
+            p
+            for m in ignored_modules
+            for p in m.parameters()
             if not isinstance(p, FlatParameter)
         )
+        # Conservatively include all shared parameters' names
         param_to_unflat_param_names = _get_param_to_unflat_param_names(
-            root_module, dedup_shared_params=False,
+            root_module,
+            dedup_shared_params=False,
         )
         ignored_param_names = set()
         for param in ignored_params:
             unflat_param_names = param_to_unflat_param_names[param]
             clean_names = []
             for k in unflat_param_names:
+                # Clean any module wrapper prefixes in case of nested wrapping
                 clean_names.append(clean_tensor_name(k))
             ignored_param_names.update(clean_names)
         return ignored_params, ignored_param_names
 
-    def _get_buffer_names(self, root_module: torch.nn.Module) -> Set[str]:
+    def _get_buffer_names(self, root_module: nn.Module) -> Set[str]:
         """
         Returns the fully prefixed names of all buffers in the module hierarchy
         rooted at ``root_module`` as a class:`set`.
-
-        Args:
-            root_module (torch.nn.Module): Top-level module passed into the
-                FSDP constructor from which to derive the fully prefixed names.
         """
-        def module_fn(module, prefix, buffer_names):
+
+        def module_fn(module: nn.Module, prefix: str, buffer_names: Set[str]):
             # For FSDP modules, only add the entry when considering the
             # contained `FlattenParamsWrapper` to avoid duplication
             if not isinstance(module, FullyShardedDataParallel):
                 for buffer_name, _ in module.named_buffers(recurse=False):
+                    # Clean module wrapper prefixes in case of nested wrapping
                     prefixed_buffer_name = clean_tensor_name(prefix + buffer_name)
                     buffer_names.add(prefixed_buffer_name)
 
-        def return_fn(buffer_names, *args):
+        def return_fn(buffer_names: Set[str], *args):
             return buffer_names
 
         buffer_names: Set[str] = set()
         return _apply_to_modules(
-            root_module, module_fn, return_fn, buffer_names,
+            root_module,
+            module_fn,
+            return_fn,
+            buffer_names,
+        )
+
+    def _auto_wrap(
```

Comment:
wonder if we can just move this to fsdp/wrap.py as a first step of decoupling this code from FSDP, since it is not really fsdp specific.

Reply:
Yeah, it may take a bit of refactoring right now, but seems mostly this function is wrapping with a wrapper cls. We'd need to do some things such as take out the assertion that checks for FSDP double wrap, and figure out how to inject the override BN for mixed precision code. After that, seems it will be more general.